File: src/stations/knx.rs
//! KNX 97.1 FM Los Angeles — segment classifier (work in progress).

use std::sync::LazyLock;

use regex::Regex;

use crate::extract::Segment;

static URL_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"\b\w+\.(com|org|net|gov|io)\b").unwrap());

static PHONE_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"\b\d{3}[-.\s]?\d{3}[-.\s]?\d{4}\b").unwrap());

static TOLLFREE_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"\b(800|888|877|866|855)\b").unwrap());

static REPORTER_SIGNOFF_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"i'm \w+ \w+.*?(k.?n?x|cbs|reporting)").unwrap());

static REPORTER_BYLINE_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"i'm \w+ \w+[,.]?\s*\w*\s*(news|fm)\b").unwrap());

fn phrase_count(blob: &str, phrases: &[&str]) -> usize {
    phrases.iter().filter(|p| blob.contains(**p)).count()
}

pub fn classify(seg: &Segment) -> String {
    let blob = seg.text_blob();
    let dur = seg.duration_secs();
    let count = seg.entry_count;

    // --- Station bumpers / IDs (short branding) ---
    const BUMPER: &[&str] = &[
        "southern california's news",
        "knx fm",
        "knx news",
        "knx hero of the week",
        "dtla law group",
    ];
    if dur < 25.0 && BUMPER.iter().any(|p| blob.contains(p)) {
        return "station_bumper".into();
    }

    // --- Station promos (longer branding segments) ---
    const PROMO: &[&str] = &[
        "you need to get out more",
        "spotlight on money",
        "grub with greg",
        "taco tuesday",
    ];
    if PROMO.iter().any(|p| blob.contains(p)) {
        return "station_promo".into();
    }

    // --- Weather forecasts ---
    const WEATHER_SIG: &[&str] = &["forecast", "meteorologist", "temperatures"];
    const WEATHER_REG: &[&str] = &["coast", "valleys", "inland empire", "mountains", "high desert"];
    if WEATHER_SIG.iter().any(|s| blob.contains(s))
        && WEATHER_REG.iter().any(|r| blob.contains(r))
    {
        return "weather".into();
    }

    // --- Market updates ---
    const MARKET: &[&str] = &["the dow", "nasdaq", "s&p", "closing bell", "money desk"];
    if phrase_count(&blob, MARKET) >= 2 {
        return "market_update".into();
    }

    // --- Podcast / show promos ---
    const PODCAST: &[&str] = &[
        "podcast",
        "wherever you get your podcasts",
        "follow and listen",
        "on demand",
        "listen on the free",
        "audible original",
    ];
    if PODCAST.iter().any(|p| blob.contains(p)) {
        return "podcast_promo".into();
    }

    // --- Ads ---
    if URL_RE.is_match(&blob) || PHONE_RE.is_match(&blob) || TOLLFREE_RE.is_match(&blob) {
        return "ad".into();
    }
    const AD: &[&str] = &[
        "call now",
        "visit us",
        "limited time",
        "free estimate",
        "free consultation",
        "free quote",
        "absolutely free",
        "no money down",
        "0% apr",
        "apr",
        "cents a day",
        "a month",
        "insurance",
        "credit union",
        "loan",
        "your home",
        "your car",
        "save money",
        "save you money",
        "special offer",
        "promo code",
        "discount",
        "schedule your",
        "book your",
        "workers' comp",
        "injury",
        "attorney",
        "termite",
        "pest",
        "timeshare",
        "exit kit",
        "download our app",
        "great clips",
        "great haircut",
        "see you in the aisles",
    ];
    if phrase_count(&blob, AD) >= 2 {
        return "ad".into();
    }

    // --- Re-aired news packages ---
    if REPORTER_SIGNOFF_RE.is_match(&blob) || REPORTER_BYLINE_RE.is_match(&blob) {
        return "news_package".into();
    }
    if count >= 3 && blob.contains("report") && dur > 20.0 {
        return "news_package".into();
    }

    // --- Expert interviews / analysis segments ---
    const INTERVIEW: &[&str] = &[
        "analyst",
        "professor",
        "expert",
        "your thoughts",
        "do you think",
        "we talked about",
        "we appreciate",
        "climate scientist",
    ];
    if phrase_count(&blob, INTERVIEW) >= 2 {
        return "interview".into();
    }

    "other".into()
}

File: repetition_analyzer/stations/knx.py
"""KNX 97.1 FM Los Angeles — segment classifier."""

import re
from typing import Callable

URL_RE = re.compile(r"\b\w+\.(com|org|net|gov|io)\b")
PHONE_RE = re.compile(r"\b\d{3}[-.\s]?\d{3}[-.\s]?\d{4}\b")
TOLLFREE_RE = re.compile(r"\b(800|888|877|866|855)\b")
REPORTER_SIGNOFF_RE = re.compile(r"i'm \w+ \w+.*?(k.?n?x|cbs|reporting)")
REPORTER_BYLINE_RE = re.compile(r"i'm \w+ \w+[,.]?\s*\w*\s*(news|fm)\b")

BUMPER = [
    "southern california's news",
    "knx fm",
    "knx news",
    "knx hero of the week",
    "dtla law group",
]

PROMO = [
    "you need to get out more",
    "spotlight on money",
    "grub with greg",
    "taco tuesday",
]

WEATHER_SIG = ["forecast", "meteorologist", "temperatures"]
WEATHER_REG = ["coast", "valleys", "inland empire", "mountains", "high desert"]

MARKET = ["the dow", "nasdaq", "s&p", "closing bell", "money desk"]

PODCAST = [
    "podcast",
    "wherever you get your podcasts",
    "follow and listen",
    "on demand",
    "listen on the free",
    "audible original",
]

AD = [
    "call now", "visit us", "limited time", "free estimate", "free consultation",
    "free quote", "absolutely free", "no money down", "0% apr", "apr",
    "cents a day", "a month", "insurance", "credit union", "loan",
    "your home", "your car", "save money", "save you money", "special offer",
    "promo code", "discount", "schedule your", "book your", "workers' comp",
    "injury", "attorney", "termite", "pest", "timeshare", "exit kit",
    "download our app", "great clips", "great haircut", "see you in the aisles",
]

INTERVIEW = [
    "analyst", "professor", "expert", "your thoughts", "do you think",
    "we talked about", "we appreciate", "climate scientist",
]


def _phrase_count(blob: str, phrases: list[str]) -> int:
    return sum(1 for p in phrases if p in blob)


def classify(seg: dict, duration_fn: Callable, text_blob_fn: Callable) -> str:
    blob = text_blob_fn(seg)
    dur = duration_fn(seg)
    count = seg.get("entry_count", 0)

    # Station bumpers / IDs
    if dur < 25.0 and any(p in blob for p in BUMPER):
        return "station_bumper"

    # Station promos
    if any(p in blob for p in PROMO):
        return "station_promo"

    # Weather
    if any(s in blob for s in WEATHER_SIG) and any(r in blob for r in WEATHER_REG):
        return "weather"

    # Market updates
    if _phrase_count(blob, MARKET) >= 2:
        return "market_update"

    # Podcast promos
    if any(p in blob for p in PODCAST):
        return "podcast_promo"

    # Ads
    if URL_RE.search(blob) or PHONE_RE.search(blob) or TOLLFREE_RE.search(blob):
        return "ad"
    if _phrase_count(blob, AD) >= 2:
        return "ad"

    # News packages
    if REPORTER_SIGNOFF_RE.search(blob) or REPORTER_BYLINE_RE.search(blob):
        return "news_package"
    if count >= 3 and "report" in blob and dur > 20.0:
        return "news_package"

    # Interviews
    if _phrase_count(blob, INTERVIEW) >= 2:
        return "interview"

    return "other"
